### PR TITLE
Feature/write dates data router

### PR DESCRIPTION
**Pull request by mmm1513**

### CODE REVIEW

File: app/backend/db/crud/date_datas.py [hunk 1]
```diff
@@ -29,11 +32,78 @@ def get_multiple_users_dates_data(
...
     start_date: date,
     end_date: date = None
-) -> List[schemas.DateResponse]:
+) -> List[schemas.RangeDatesResponse]:
+
     return [
         {
             'user_id': user_id,
             'data': get_dates_data(db, user_id, start_date, end_date)
         }
```
Comment (IamShobe) [inline app/backend/db/crud/date_datas.py]: isnt it better to return a dict of key = user_id, value = get_dates_data?

File: app/backend/db/crud/date_datas.py [hunk 2]
```diff
@@ -29,11 +32,78 @@ def get_multiple_users_dates_data(
...
         for user_id in users_id
     ]
+
+
+def set_new_date_data(
+    db: Session, 
+    body=schemas.PostDateDataBody
```
Comment (IamShobe) [inline app/backend/db/crud/date_datas.py]: in crud i dont want to see parameter as body...
i prefer u to expand it here -
start_date
end_date
reason

File: app/backend/db/crud/date_datas.py [hunk 3]
```diff
@@ -29,11 +32,78 @@ def get_multiple_users_dates_data(
...
+) -> schemas.RangeDatesResponse:
+
+    dates_data = []
+    for date in daterange(body.start_date, body.end_date):
+        if reason is not None:
```
Comment (IamShobe) [inline app/backend/db/crud/date_datas.py]: wat if reason is None?

Comment (IamShobe) [reply app/backend/db/crud/date_datas.py]: we should limit it i think
or atleast raise an exception

File: app/backend/db/crud/date_datas.py [hunk 4]
```diff
@@ -29,11 +32,78 @@ def get_multiple_users_dates_data(
...
+    for user_id in users_id:
+        dates_data_to_remove = get_dates_data(
+                db=db, 
+                user_id=user_id, 
+                start_date=start_date, 
+                end_date=end_date
+            )
+
+        for date_data in dates_data_to_remove:
+            db.delete(date_data)
```
Comment (IamShobe) [inline app/backend/db/crud/date_datas.py]: u forgot to commit

File: app/backend/db/crud/date_datas.py [hunk 5]
```diff
@@ -29,11 +32,78 @@ def get_multiple_users_dates_data(
...
+
+def put_data_in_user(
+    db: Session, 
+    body=schemas.PutDateDataBody
```
Comment (IamShobe) [inline app/backend/db/crud/date_datas.py]: expand parameters..

File: app/backend/db/crud/date_datas.py [hunk 6]
```diff
@@ -29,11 +32,78 @@ def get_multiple_users_dates_data(
...
+) -> schemas.RangeDatesResponse:
+
+    request = body.dict()
+    if body.reason is not None:
```
Comment (IamShobe) [inline app/backend/db/crud/date_datas.py]: what is reason None?

File: app/backend/db/crud/date_datas.py [hunk 7]
```diff
@@ -29,11 +32,78 @@ def get_multiple_users_dates_data(
...
+            db=db, 
+            user_id=request.pop("user_id"),
+            start_date=request.pop("start_date"),
+            end_date=request.pop("end_date")
+        )
+
+    for datedata in dates_data:
+        for key, value in body.dict().items():
+            if value is not None:
+                setattr(datedata, key, value)
```
Comment (IamShobe) [inline app/backend/db/crud/date_datas.py]: What??
please dont..

File: app/backend/db/crud/date_datas.py [hunk 8]
```diff
@@ -29,11 +32,78 @@ def get_multiple_users_dates_data(
...
+
+    request = body.dict()
+    if body.reason is not None:
+        request["reason"] = get_reason_by_name(db, body.reason)
+
+    dates_data = get_dates_data(
+            db=db, 
+            user_id=request.pop("user_id"),
+            start_date=request.pop("start_date"),
+            end_date=request.pop("end_date")
```
Comment (IamShobe) [inline app/backend/db/crud/date_datas.py]: ....

File: app/backend/server/api/v1/dates_router.py [hunk 9]
```diff
@@ -26,6 +23,35 @@
...
                                               users_id=users_id)
 
+@router.post("/", response_model=schemas.RangeDatesResponse)
+async def post_dates_status(
+    body: schemas.PostDateDataBody,
+    db: Session = Depends(get_db),
+    current_user: schemas.User = Security(get_current_user,
+                                          scopes=["personal"])
+):
+    return crud.set_new_date_data(db=db, body=body)
```
Comment (IamShobe) [inline app/backend/server/api/v1/dates_router.py]: expand body here

File: app/backend/server/api/v1/dates_router.py [hunk 10]
```diff
@@ -26,6 +23,35 @@
                                               end_date=end,
                                               users_id=users_id)
 
+@router.post("/", response_model=schemas.RangeDatesResponse)
+async def post_dates_status(
+    body: schemas.PostDateDataBody,
```
Comment (IamShobe) [inline app/backend/server/api/v1/dates_router.py]: y ...
expand it here also
its ugly like that
i want to see here right away what params to look for

File: app/backend/server/api/v1/dates_router.py [hunk 11]
```diff
@@ -26,6 +23,35 @@
...
+                            users_id=users_id)
+
+@router.put("/", response_model=schemas.RangeDatesResponse)
+async def put_dates_status(
+    body: schemas.PutDateDataBody,
+    db: Session = Depends(get_db),
+    current_user: schemas.User = Security(get_current_user,
+                                          scopes=["personal"])
+):
+    return crud.put_data_in_user(db=db, body=body)
```
Comment (IamShobe) [inline app/backend/server/api/v1/dates_router.py]: expand body here too

Comment (IamShobe) [reply app/backend/server/api/v1/dates_router.py]: get rid of the PutDateDataBody here and everywhere

File: app/backend/db/crud/date_datas.py (hunk 1 above)
Comment (mmm1513) [reply app/backend/db/crud/date_datas.py]: IDK I don't mind

File: app/backend/db/crud/date_datas.py (hunk 3 above)
Comment (mmm1513) [reply app/backend/db/crud/date_datas.py]: So raise exception only if it's None - and status is here

File: app/backend/db/crud/date_datas.py (hunk 6 above)
Comment (mmm1513) [reply app/backend/db/crud/date_datas.py]: it's put, you don't have to put everything

File: app/backend/db/crud/date_datas.py (hunk 7 above)
Comment (mmm1513) [reply app/backend/db/crud/date_datas.py]: Why? 😂

File: app/backend/db/crud/date_datas.py (hunk 8 above)
Comment (mmm1513) [reply app/backend/db/crud/date_datas.py]: 3 points is not something I can resolve

File: app/backend/db/crud/date_datas.py (hunk 7 above)
Comment (IamShobe) [reply app/backend/db/crud/date_datas.py]: dynamic programming is really bad...
i will be hell to debug later

File: app/backend/db/crud/date_datas.py (hunk 8 above)
Comment (IamShobe) [reply app/backend/db/crud/date_datas.py]: just expand the body and dont use request

File: app/backend/db/crud/date_datas.py (hunk 1 above)
Comment (mmm1513) [reply app/backend/db/crud/date_datas.py]: done

File: app/backend/db/crud/date_datas.py (hunk 2 above)
Comment (mmm1513) [reply app/backend/db/crud/date_datas.py]: done

File: app/backend/db/crud/date_datas.py (hunk 3 above)
Comment (mmm1513) [reply app/backend/db/crud/date_datas.py]: done

File: app/backend/db/crud/date_datas.py (hunk 4 above)
Comment (mmm1513) [reply app/backend/db/crud/date_datas.py]: done

File: app/backend/db/crud/date_datas.py (hunk 5 above)
Comment (mmm1513) [reply app/backend/db/crud/date_datas.py]: done

File: app/backend/db/crud/date_datas.py (hunk 7 above)
Comment (mmm1513) [reply app/backend/db/crud/date_datas.py]: done

File: app/backend/db/crud/date_datas.py (hunk 8 above)
Comment (mmm1513) [reply app/backend/db/crud/date_datas.py]: done

File: app/backend/server/api/v1/dates_router.py (hunk 9 above)
Comment (mmm1513) [reply app/backend/server/api/v1/dates_router.py]: done

File: app/backend/server/api/v1/dates_router.py (hunk 10 above)
Comment (mmm1513) [reply app/backend/server/api/v1/dates_router.py]: whtat?! this is not how this works :o

File: app/backend/server/api/v1/dates_router.py (hunk 11 above)
Comment (mmm1513) [reply app/backend/server/api/v1/dates_router.py]: done in the expand part

File: app/backend/db/crud/date_datas.py [hunk 12]
```diff
@@ -44,26 +44,36 @@ def get_multiple_users_dates_data(
...
-    db: Session, 
-    body=schemas.PostDateDataBody
+    db: Session,
+    user_id: int,
+    start_date: date,
+    end_date: date = None,
+    state: schemas.AnswerStateTypes = None,
+    reason: str = None,
+    reported_by_id: int = None,
+    reported_time: datetime = None
```
Comment (IamShobe) [inline app/backend/db/crud/date_datas.py]: its so much better !

Comment (IamShobe) [reply app/backend/db/crud/date_datas.py]: maybe do the reported_time required?

Comment (IamShobe) [reply app/backend/db/crud/date_datas.py]: or default it to time.time()

Comment (IamShobe) [reply app/backend/db/crud/date_datas.py]: state also must be mendatory

File: app/backend/db/crud/date_datas.py [hunk 13]
```diff
@@ -44,26 +44,36 @@ def get_multiple_users_dates_data(
...
-                state=body.state, 
+                user_id=user_id,
+                state=state, 
                 reason=reason, 
-                reported_by=body.reported_by, 
-                reported_time=body.reported_time
+                reported_by=reported_by, 
+                reported_time=reported_time
             ))
     db.add_all(dates_data)
```
Comment (IamShobe) [inline app/backend/db/crud/date_datas.py]: ok you got one thing wrong is that u forgot to actually create DateDetails instance if not exists

File: app/backend/db/crud/date_datas.py [hunk 14]
```diff
@@ -83,27 +93,42 @@ def delete_users_dates_data(
...
+    reported_by_id: int = None,
+    reported_time: datetime = None
 ) -> schemas.RangeDatesResponse:
 
-    request = body.dict()
-    if body.reason is not None:
-        request["reason"] = get_reason_by_name(db, body.reason)
+
+    if reason is not None:
+        reason = get_reason_by_name(db, reason)
```
Comment (IamShobe) [inline app/backend/db/crud/date_datas.py]: just do the reason must
dont allow default value

Comment (IamShobe) [reply app/backend/db/crud/date_datas.py]: here and everywhere

File: app/backend/db/crud/date_datas.py [hunk 15]
```diff
@@ -83,27 +93,42 @@ def delete_users_dates_data(
...
+            datedata.state = state
+
+        if reason is not None:
+            datedata.reason = reason
+
+        if reported_by_id is not None:
+            datedata.reported_by_id = reported_by_id
+
+        if reported_time is not None:
+            datedata.reported_time = reported_time
```
Comment (IamShobe) [inline app/backend/db/crud/date_datas.py]: cant u do it just with .update function??
just do a list cleanups of all None

File: app/backend/server/api/v1/dates_router.py [hunk 16]
```diff
@@ -51,7 +58,15 @@
...
 ):
-    return crud.put_data_in_user(db=db, body=body)
+    return crud.put_data_in_user(
+            db=db, 
+            start_date=body.start_date,
+            end_date=body.end_date,
+            state=body.state,
+            reason=body.reason,
+            reported_by_id=body.reported_by_id,
+            reported_time=body.reported_time
```
Comment (IamShobe) [inline app/backend/server/api/v1/dates_router.py]: i still dont like the usage of body...
but ok leave it..

File: app/backend/db/crud/date_datas.py (hunk 12 above)
Comment (mmm1513) [reply app/backend/db/crud/date_datas.py]: My bad, I've copied from put

File: app/backend/db/crud/date_datas.py (hunk 14 above)
Comment (mmm1513) [reply app/backend/db/crud/date_datas.py]: Here is put, I want to allow a change of a single value without have to give the rest

File: app/backend/db/crud/date_datas.py (hunk 12 above)
Comment (mmm1513) [reply app/backend/db/crud/date_datas.py]: done

File: app/backend/db/crud/date_datas.py (hunk 13 above)
Comment (mmm1513) [reply app/backend/db/crud/date_datas.py]: i thought it makes one automatically :P 
done.

File: app/backend/db/crud/date_datas.py (hunk 15 above)
Comment (mmm1513) [reply app/backend/db/crud/date_datas.py]: i wanted but you told me not to..
image

File: app/backend/db/models/date_datas.py [hunk 17]
```diff
@@ -24,7 +33,7 @@ class DateData(Base):
     date_details = relationship('DateDetails', backref='datas')
     user_id = Column(Integer, ForeignKey('users.id', ondelete='CASCADE'), index=True)
     user = relationship('User', backref='statuses', foreign_keys=[user_id])
-    state = Column(Enum('here', 'not_here', name='answer_state'))
+    state = Column(Enum(Reason.HERE, Reason.NOT_HERE, name='answer_state'))
```
Comment (IamShobe) [inline app/backend/db/models/date_datas.py]: You changed the model..
Run alembic to add migration
Make sure it has changes the enum values and default value ( I know it have problem with it)

File: app/backend/db/crud/crud_utils.py [hunk 18]
```diff
@@ -0,0 +1,6 @@
+def put_values_if_not_none(db, obj, **kwargs):
+    for attr, value in kwargs.items():
+        if value is not None:
+            setattr(obj, attr, value)
+    
+    db.commit()
```
Comment (IamShobe) [inline app/backend/db/crud/crud_utils.py]: Add refresh to obj after.
Add new line at the end

Comment (mmm1513) [reply app/backend/db/crud/crud_utils.py]: done

File: app/backend/db/models/date_datas.py (hunk 17 above)
Comment (mmm1513) [reply app/backend/db/models/date_datas.py]: done, didnt need to change anything

**Comment by IamShobe** [comment]
please add tests :(

File: Pipfile [hunk 19]
```diff
@@ -27,6 +27,8 @@ pyjwt = "*"
 python-multipart = "*"
 gunicorn = "*"
 psycopg2-binary = "*"
+boltons = "*"
```
Comment (IamShobe) [inline Pipfile]: delete

File: Pipfile [hunk 20]
```diff
@@ -27,6 +27,8 @@ pyjwt = "*"
...
+pytest = "*"
```
Comment (IamShobe) [inline Pipfile]: delete..
pytest is dev dependency

File: app/backend/db/crud/reasons.py [hunk 21]
```diff
@@ -1,12 +1,20 @@
...
 
 from db.models import Reason
 
 
-def get_reasons(db: Session) -> List[Reason]:
+def get_reason_by_name(db: Session, reason: str) -> str:
+    """Get reason object from name."""
+    reason_to_return = db.query(Reason).filter(Reason.name == reason).first()
+    if reason_to_return is None:
+        raise AttributeError(f"Reason: {reason} does not exsists in the DB.")
```
Comment (IamShobe) [inline app/backend/db/crud/reasons.py]: ValueError

File: app/backend/server/auth.py [hunk 22]
```diff
@@ -164,7 +164,7 @@ def create_access_token(*, data: dict, expires_delta: timedelta = None) -> str:
 @router.post("/register", response_model=User)
 def register(
     *,
-    username: str = Body(..., regex=r"^[A-Za-z0-9]+$"),
+    username: str = Query(..., regex=r"^[A-Za-z0-9]+$"),
```
Comment (IamShobe) [inline app/backend/server/auth.py]: y did u change it to query>?

**Comment by IamShobe** [comment]
fix pylint

File: .env [hunk 23]
```diff
@@ -1 +1 @@
-PYTHONPATH=${PWD}/app/backend
+PYTHONPATH=${PWD}/app/backend:${PWD}/tests
```
Comment (IamShobe) [inline .env]: delete from env

**Pull request merged**